app/src/main/java/com/graeberj/taskman/auth/presentation/components/TaskTextField.kt

```diff
@@ -27,11 +50,101 @@ fun CustomTextField(
+        ),
+        colors = TextFieldDefaults.textFieldColors(
+            focusedTrailingIconColor = if (isValid) TaskyGreen else trailingIconColor ?: Color.Gray
+        ),
+        maxLines = maxLines,
+        singleLine = maxLines == 1,
         trailingIcon = {
-            Icon(imageVector = Icons.Default.Check, contentDescription = "check")
+            TrailingIcon(
+                isPassword = isPassword,
+                trailingIcon = trailingIcon,
+                onClick = onClick,
+                isValid = isValid,
+                isHidden = isHidden
+            )
         },
-        shape = RoundedCornerShape(10.dp)
+        shape = RoundedCornerShape(10.dp),
+        visualTransformation = if (isHidden) PasswordVisualTransformation()
+        else VisualTransformation.None
     )
+}
+
+@Composable
+fun EmailTextField(
+    modifier: Modifier = Modifier,
+    value: String,
+    onValueChange: (String) -> Unit,
+    placeholder: String = "",
+    shouldShowError: Boolean = false,
+    isValid: Boolean = false,
+) {
+    CustomTextField(
+        value = value,
+        onValueChange = onValueChange,
+        shouldShowError = shouldShowError,
+        placeholder = placeholder,
+        modifier = modifier,
+        keyboardType = KeyboardType.Email,
+        isValid = isValid
+    )
+}
+
+@Composable
+fun PasswordTextField(
+    modifier: Modifier = Modifier,
+    value: String,
+    onValueChange: (String) -> Unit,
+    placeholder: String = "",
+    shouldShowError: Boolean = false,
+    isValid: Boolean = false,
+    onIconClicked: () -> Unit,
+    isHidden: Boolean = true,
+) {
+    CustomTextField(
+        value = value,
+        onValueChange = onValueChange,
+        shouldShowError = shouldShowError,
+        placeholder = placeholder,
+        modifier = modifier,
+        keyboardType = KeyboardType.Password,
+        isValid = isValid,
+        isPassword = true,
+        onClick = onIconClicked,
+        isHidden = isHidden,
+    )
+}
+@Composable
+private fun TrailingIcon(
+    isPassword: Boolean = false,
+    isHidden: Boolean = false,
+    trailingIcon: ImageVector? = null,
+    isValid: Boolean = false,
+    onClick: () -> Unit = {}
+) {
+    if (trailingIcon != null) {
+        IconButton(onClick = onClick) {
+            Icon(
+                imageVector = trailingIcon,
+                contentDescription = stringResource(id = R.string.trailing_icon)
+            )
+        }
+    } else if (isPassword) {
+        IconButton(onClick = onClick) {
+            Icon(
+                imageVector = if (isHidden) Icons.Default.VisibilityOff
+                else Icons.Default.Visibility,
+                contentDescription = stringResource(R.string.toggle_password_visibility)
+            )
+        }
+    } else if (isValid) {
+        Icon(
+            imageVector = Icons.Default.Check,
+            contentDescription = stringResource(R.string.input_is_valid)
+        )
+    }
```

Comment:
Matter of taste, but for so many different branches, I personally find a `when` block more readable